include/boost/type_index/runtime_cast.hpp

```diff
@@ -0,0 +1,141 @@
+    }
+
+    template<class T, class FirstBase, class... Rest>
+    void* check_bases(T* p, type_index const& idx) const BOOST_NOEXCEPT {
+        if(void* result = p->FirstBase::boost_type_index_find_instance_(idx))
+            return result;                                      
+        return check_bases<T, Rest...>(p, idx); 
+    }
+
+    template<class T>
+    void* operator()(T* p, type_index const& idx) const BOOST_NOEXCEPT {
+        if(Current* current = check_current(p, idx))        
+            return p;                                               
+        return check_bases<T, BaseList...>(p, idx);
+    }
+};
+
+template<typename T, typename U>
+T* runtime_cast_impl(U* u, std::true_type) {
+    return u;
+} 
+
+template<typename T, typename U>
+T const* runtime_cast_impl(U const* u, std::true_type) {
+    return u;
+}
+
+template<typename T, typename U>
+T* runtime_cast_impl(U* u, std::false_type) {
+    return static_cast<T*>(
+        u->boost_type_index_find_instance_(boost::typeindex::type_id<T>())
+    );
+} 
+
+template<typename T, typename U>
+T const* runtime_cast_impl(U const* u, std::false_type) {
+    return static_cast<T const*>(
+        const_cast<U*>(u)->boost_type_index_find_instance_(boost::typeindex::type_id<T>())
+    );
+}
+
+} // namespace detail
+
+#define BOOST_TYPE_INDEX_REGISTER_CLASS_RTTI                                                                \
+    virtual void* boost_type_index_find_instance_(boost::typeindex::type_index const& idx) BOOST_NOEXCEPT { \
+        if(idx == boost::typeindex::type_id<std::decay<decltype(*this)>::type>())                           \
+            return this;                                                                                    \
+        return nullptr;                                                                                     \
+    }                                                                   
+
+#define BOOST_TYPE_INDEX_REGISTER_CLASS_RTTI_BASES(...)                                                         \
+    virtual void* boost_type_index_find_instance_(boost::typeindex::type_index const& idx) BOOST_NOEXCEPT {     \
+        return boost::typeindex::detail::find_type<std::decay<decltype(*this)>::type, __VA_ARGS__>()(this, idx);\
+    }  
+
+    template<typename T, typename U>
+    T runtime_cast(U* u) BOOST_NOEXCEPT {
+        typedef typename std::remove_pointer<T>::type impl_type;
+        return detail::runtime_cast_impl<impl_type>(u, std::is_same<T, U>());
+    } 
+
+    template<typename T, typename U>
+    T runtime_cast(U const* u) BOOST_NOEXCEPT {
+        typedef typename std::remove_pointer<T>::type impl_type;
+        return detail::runtime_cast_impl<impl_type>(u, std::is_same<T, U>());
+    }
+
+    template<typename T, typename U>
+    T runtime_cast(U& u) {
+        typedef typename std::remove_reference<T>::type impl_type;
+        impl_type* value = detail::runtime_cast_impl<impl_type>(&u, std::is_same<T, U>());
```

Comment:
It's better to use boost::is_same, because it works even on antique compilers and c++98 compatible standard libraries